File: ios/Sources/TaigiKeyboard/Lexicon/Utils/LexiconBitmask.swift
// 把 dictionary.bin 的 record bitmask 解碼成 [DictionarySource]。
// bit 位置由 dictionary/common/source_bits.py 與 Android LexiconBitmask 共同擁有,
// 三邊不能漂移,否則 filter / ranking 會靜默分歧。

import Foundation

/// `LexiconBitmask` — bitmask → `[DictionarySource]` decoder.
///
/// CROSS-PLATFORM INVARIANT — bit positions mirror
/// `dictionary/common/source_bits.py` (SOURCE_BITS + IS_VARIANT_BIT at
/// bit 12), `engine/lexicon/src/dictionary_reader.rs` constants, and
/// Android `LexiconBitmask` equivalent. Drift causes silent filter +
/// ranking divergence.
enum LexiconBitmask {
    /// Decode a record bitmask (`UInt32` from `RustEngineBridge.LexiconRow`)
    /// into the ordered list of `DictionarySource` values. Order matches
    /// the bit position so iOS UI badges render predictably.
    // 解碼 record bitmask → [DictionarySource]。
    // 順序跟著 bit 位置走,讓 UI badge 渲染有可預測順序。
    static func sources(from bitmask: UInt32) -> [DictionarySource] {
        let pairs: [(bit: UInt32, source: DictionarySource)] = [
            (1 << 0, .kautian), (1 << 1, .taigitv), (1 << 2, .itaigi), (1 << 3, .sitbut),
            (1 << 4, .taihoa), (1 << 5, .taijit), (1 << 6, .kungge), (1 << 7, .stti),
            (1 << 8, .khpoo), (1 << 9, .khiin), (1 << 10, .dev), (1 << 11, .lkk),
        ]
        return pairs.compactMap { bitmask & $0.bit != 0 ? $0.source : nil }
    }
}

File: android/app/src/main/java/com/siansiansu/taigikeyboard/ime/dictionary/LexiconBitmask.kt
package com.siansiansu.taigikeyboard.ime.dictionary

/**
 * `LexiconBitmask` — bitmask → `List<DictionarySource>` decoder.
 *
 * CROSS-PLATFORM INVARIANT — bit positions mirror
 * `dictionary/common/source_bits.py` (SOURCE_BITS + IS_VARIANT_BIT at
 * bit 12), `engine/lexicon/src/dictionary_reader.rs` constants, and iOS
 * `LexiconBitmask`. Drift causes silent filter + ranking divergence.
 */
object LexiconBitmask {
    fun sourcesFromBitmask(bitmask: Int): List<DictionarySource> {
        val pairs = listOf(
            (1 shl 0) to DictionarySource.KAUTIAN,
            (1 shl 1) to DictionarySource.TAIGITV,
            (1 shl 2) to DictionarySource.ITAIGI,
            (1 shl 3) to DictionarySource.SITBUT,
            (1 shl 4) to DictionarySource.TAIHOA,
            (1 shl 5) to DictionarySource.TAIJIT,
            (1 shl 6) to DictionarySource.KUNGGE,
            (1 shl 7) to DictionarySource.STTI,
            (1 shl 8) to DictionarySource.KHPOO,
            (1 shl 9) to DictionarySource.KHIIN,
            (1 shl 10) to DictionarySource.DEV,
            (1 shl 11) to DictionarySource.LKK,
        )
        return pairs.mapNotNull { (bit, source) ->
            if (bitmask and bit != 0) source else null
        }
    }
}
